src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewFactory.cs

```diff
@@ -0,0 +1,65 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Responsible for wiring up all the different components of the document outline feature.
+    /// </summary>
+    internal static class DocumentOutlineViewFactory
+    {
+        public static DocumentOutlineView CreateView(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            threadingContext.ThrowIfNotOnUIThread();
+            var (textViewEventSource, textBuffer) = CreateEventSource(asyncListener, editorAdaptersFactoryService, codeWindow);
+            var viewModel = new DocumentOutlineViewModel(languageServiceBroker, asyncListener, textViewEventSource, textBuffer, threadingContext);
+            return new DocumentOutlineView(viewModel, editorAdaptersFactoryService, codeWindow);
+        }
+
+        private static (CompilationAvailableTaggerEventSource, ITextBuffer) CreateEventSource(
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            if (ErrorHandler.Failed(codeWindow.GetLastActiveView(out var textView)))
+            {
+                FailFast.Fail("Unable to get the last active text view. IVsCodeWindow implementation we are given is invalid.");
+            }
+
+            var wpfTextView = editorAdaptersFactoryService.GetWpfTextView(textView);
+            Assumes.NotNull(wpfTextView);
+
+            var subjectBuffer = wpfTextView.TextBuffer;
+            var textViewEventSource = new CompilationAvailableTaggerEventSource(
+                subjectBuffer,
+                asyncListener,
+                // Any time an edit happens, recompute as the document symbols may have changed.
+                TaggerEventSources.OnTextChanged(subjectBuffer),
+                // Switching what is the active context may change the document symbols.
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer),
+                // Many workspace changes may need us to change the document symbols (like options changing, or project renaming).
```

Comment:
This comment seems odd, since I can't imagine what options changes or project renames would change the list of document symbols...?

Reply:
This is a comment I copy/pasted without much thought. I'll investigate and update.

Reply:
So, compilation options changes could change the status of parsing. mainly, if a compiler constant is/is not defined the `#ifdef` would parse differently giving us a different tree. `OnDocumentActiveContextChanged` is not necessary though, as we are disposed/recreated in this case.

Reply:
can we remove that then?

Reply:
it has been removed